controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java

```diff
@@ -1686,23 +1696,41 @@ private StreamSegmentRecord newSegmentRecord(int epoch, int segmentNumber, long
     // endregion
 
     // region transactions
-    abstract CompletableFuture<Version> createNewTransaction(final int epoch, final UUID txId, final byte[] record);
+    abstract CompletableFuture<Version> createNewTransaction(final UUID txId, final byte[] record);
 
-    abstract CompletableFuture<Data> getActiveTx(final int epoch, final UUID txId);
+    abstract CompletableFuture<Data> getActiveTx(final UUID txId);
 
-    abstract CompletableFuture<Version> updateActiveTx(final int epoch,
-                                                       final UUID txId,
-                                                       final Data data);
+    abstract CompletableFuture<Version> updateActiveTx(final UUID txId, final Data data);
+
+    // This method is called for adding a transcation to committing order. 
+    // It is important to note that this method is called before marking transaction as committing in active txn record. 
+    // So retrieving a transaction from this ordered list is no guarantee that the transaction has been set to committing. 
+    // Similarly, there could be duplicate entries for the same transaction if retried. 
+    // So the ordered ist merely captures the order in which a request may have been received and is only opportunistic 
+    // and makes no strong claim of its consistency with active txn record. 
+    // 1. we could have duplicate entries for same transaction in commit order.
+    // 2. we could have transcations that are marked for aborting added to commit order
+    // 3. we could have transactions that are no longer in activeTxn list yet present in commit order. 
+    abstract CompletableFuture<Long> addTxnToCommitOrder(final UUID txId);
+
+    abstract CompletableFuture<Void> removeTxnsFromCommitOrder(final List<Long> positions);
 
     abstract CompletableFuture<Data> getCompletedTx(final UUID txId);
 
-    abstract CompletableFuture<Void> removeActiveTxEntry(final int epoch, final UUID txId);
+    abstract CompletableFuture<Void> removeActiveTxEntry(final UUID txId);
 
     abstract CompletableFuture<Void> createCompletedTxEntry(final UUID txId, byte[] data);
 
     abstract CompletableFuture<Map<String, Data>> getCurrentTxns();
 
-    abstract CompletableFuture<Map<String, Data>> getTxnInEpoch(int epoch);
+    /**
+     * This method finds transactions to commit in lowest epoch and returns a sorted list of transaction ids, 
+     * sorted by their order of commits. 
+     */
```

Comment:
I'm sure this will introduce a new Javadoc warning due to missing `@return` tag.